rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,510 @@
+  process, allowing use of containers, secret stores, and other security
+  measures.
+
+Here is a sample config that connects to an HTTP-based MCP server:
+```yaml
+app_service:
+  enabled: true
+  apps:
+    - name: "everything"
+      uri: "mcp+http://localhost:3001/mcp"
+```
+
+`tsh mcp connect` by default serves stdio transport to the AI client tool but
+translates stdio to streamable HTTP before sending it to Teleport backend.
+
+The MCP server handler on the App service will run an in-memory HTTP server that
+handles the MCP protocol before forwarding the requests to the URI specified in
+the App definition.
+
+### Phase 2 - expired tsh session handling
+
+MCP server errors are permanent in Claude Desktop. When `tsh` session expires,
+the user has to run `tsh login` then restart Claude Desktop.
+
+<img src="assets/0209-claude-desktop-disconnected.png" alt="Claude Desktop disconnecting" height="100"/>
+
+To properly handle this, a mini JSON-RPC server should be served locally that
+ensures a good connection with Claude Desktop. When `tsh` session expires, the
+local JSON-RPC should instruct the user to perform a `tsh login` while
+re-establish the connection with Proxy when possible. Note that the local
+JSON-RPC server does NOT need to understand/implement the full MCP protocol as
+it only forwards the protocol to backend when things are good and rejects the
+request with hint when `tsh` session expires.
+
+### Phase 2 - "tools/list" filtering
+
+The initial implementation can deny access to specific tools through
+"tools/call". However, users can still see all available tools from the MCP
+server, including the ones they don't have access to.
+
+Ideally, the response of "tools/list" should be filtered so only allowed tools
+are present. However, supporting this filter requires a more complex server
+implementation and a lot driver refactoring. Special handling on
+"notifications/tools/list_changed" may also be necessary.
+
+### Other improvements - RBAC on resources and prompts
+
+Currently, MCP protocol supports three primitives: resources, tools, and
+prompts. We can extend `role.allow.mcp` for primitives other than tools:
+
+```yaml
+allow:
+  mcp:
+    tools:
+    - "*"
+    resources:
+    - "file:///var/log/*"
+    prompts:
+    - "code_review"
+```
+
+Note that both resources and prompts support
+[Completion](https://modelcontextprotocol.io/specification/2025-03-26/server/utilities/completion),
+which might also need some handling.
+
+### Other improvements - OAuth support
+
+Teleport can support [OAuth
+authorization](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization)
+between Teleport service and the MCP servers using the "Client Credentials"
+grant type. Client ID and secrets can potentially be saved as integration
+credentials.
+
+However, the OAuth support has a dependency on non-stdio transport. More
+importantly, we should monitor adoption trends to determine whether OAuth
+support is necessary.
```

Comment:
I was reading the [MCP OAuth spec](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization) and thinking that it might actually be more useful if Teleport itself could act as an OAuth IdP for MCP servers supporting it. Then it would be similar to us providing access to SAML apps for example via our SAML IdP. What do you think?

Reply:
> Teleport itself could act as an OAuth IdP for MCP servers supporting it

I've thought about it, though in that case, would the MCP protocol still be proxied through Teleport or the client connects directly to the MCP server? if not proxied, i feel it's more an "identity" feature than "access", and we wouldn't be able to apply our RBAC system and do audit logging.

On the other hand, it may also be very useful if Teleport itself act as OAuth Idp for MCP servers NOT supporting it. That case Teleport will host HTTP-based MCP servers at a public Proxy endpoint and clients connect directly to Proxy without `tsh`.  @gabrielcorado shared this today and seems Anthropic is pushing this way:
https://www.anthropic.com/news/integrations

Reply:
updated the oauth section with all three possible expansions